src/backend/joanie/payment/models.py

```diff
@@ -1 +1,463 @@
+        _("invoice address"), max_length=255, null=False, blank=False
+    )
+    localized_context = models.JSONField(
+        _("context"),
+        help_text=_("Localized data that needs to be frozen on invoice creation"),
+        editable=False,
+    )
+
+    class Meta:
+        db_table = "joanie_invoice"
+        verbose_name = _("Invoice")
+        verbose_name_plural = _("Invoices")
+        constraints = [
+            models.CheckConstraint(
+                check=(models.Q(order__isnull=False) & models.Q(parent__isnull=True))
+                | (models.Q(order__isnull=True) & models.Q(parent__isnull=False)),
+                name="invoice_relies_on_order_or_invoice",
+            ),
+        ]
+
+    def __str__(self):
+        types = dict(payment_enums.INVOICE_TYPES)
+        return f"{types[self.type]} {self.reference}"
+
+    @property
+    def root_parent(self):
+        """
+        Return the root parent invoice.
+        """
+        parent = self
+
+        while parent.parent is not None:
+            parent = parent.parent
+
+        return parent
+
+    @property
+    def state(self):
+        """
+        Process the state of the invoice
+        """
+        if self.balance.amount >= 0:
+            if (
+                self.invoiced_balance.amount == 0
+                and self.transactions_balance.amount == 0
+            ):
+                return payment_enums.INVOICE_STATE_REFUNDED
+
+            return payment_enums.INVOICE_STATE_PAID
+
+        return payment_enums.INVOICE_STATE_UNPAID
+
+    def get_related_order(self):
+        """
+        Return the related order.
+        """
+        return self.root_parent.order
+
+    @property
+    def document(self):
+        """
+        Get the document related to the invoice instance;
+        """
+        # pylint: disable=no-member
+        document = InvoiceDocument(context_query=self.get_document_context())
+        return document.create(persist=False)
+
+    @property
+    def type(self):
+        """
+        Return the invoice type according to its amount.
+        If amount is positive, invoice type is "invoice"
+        otherwise it's "credit_note"
+        """
+        if self.amount.amount < 0:  # pylint: disable=no-member
+            return payment_enums.INVOICE_TYPE_CREDIT_NOTE
+
+        return payment_enums.INVOICE_TYPE_INVOICE
+
+    @cached_property
+    def all_children(self):
+        """
+        Return all children invoices.
+        """
+        # - Retrieve invoice and children invoice data
+        query = """
+        WITH RECURSIVE parents AS (
+            SELECT invoice.*
+            FROM joanie_invoice invoice
+            WHERE id = %s
+
+            UNION ALL
+
+            SELECT joanie_invoice.*
+            FROM joanie_invoice,parents
+            WHERE joanie_invoice.parent_id = parents.id
+        )
+        SELECT * FROM parents WHERE id != %s
+        """
+
+        return Invoice.objects.raw(query, [self.id, self.id])
```

Comment:
maybe with the additional constraint I proposed you can do:
```python
def related_invoices(self):
    return Invoice.objects.filter(Q(id=self.id) | Q(parent__id=self.id))
```

Reply:
As child depth is limited to 1 level, we are now able to use directly `self.children.all()`.